### DISCUSSION TITLE
Impossible to install Tensorflow on dockerized Airflow 2.3.2

### DISCUSSION THREAD

**Discussion by Victor-Gordaliza-VM** [apache/airflow#24920] [Q&A] [answered]
I'm trying to update to Airflow 2.3.2 and I need to install Tensorflow due some small part of the code needed.

**requirements.txt**
```
airflow-multi-dagrun==2.2.0
apache-airflow-providers-amazon
apache-airflow-providers-celery
apache-airflow-providers-docker
apache-airflow-providers-postgres
apache-airflow-providers-redis
docker
psycopg2-binary
tensorflow==2.9
```
**Dockerfile** installing the requirements with the *constraint* flag
```
FROM apache/airflow:slim-2.3.2-python3.10

# Install deps as root.
USER root

RUN apt-get update \
    && apt-get install -y git wget build-essential libsndfile1 --no-install-recommends \
    && apt-get autoclean -y && apt-get clean -y 

USER airflow
# Install python requirements
RUN /usr/local/bin/python3 -m pip install --upgrade pip
ADD requirements.txt /requirements.txt
RUN pip install -r /requirements.txt --no-cache-dir --constraint "https://raw.githubusercontent.com/apache/airflow/constraints-2.3.2/constraints-3.10.txt"
```

When I try to create the image, the next error appear:
```
ERROR: Cannot install tensorflow because these package versions have conflicting dependencies.

The conflict is caused by:
    tensorboard 2.9.1 depends on google-auth-oauthlib<0.5 and >=0.4.1
    tensorboard 2.9.0 depends on google-auth-oauthlib<0.5 and >=0.4.1
    The user requested (constraint) google-auth-oauthlib==0.5.1
```

I've tried with Tensorflow 2.8 but the same error raises. 

Since the use of "constraints" is almost mandatory, is there a solution to this? For our part we are looking to get the functionality out to an external container, but I would like to know if there is a solution.

Thanks!

**Comment by potiuk** [comment, marked as answer]
There is nothing we can do. You should open issue in tensorboard to upgrade to newer version of the google-auth-oauthlib dependency.


However you are not  blocked. You are totally mistaken about the reason why we have constraints. Constraints are not mandatory. They are there to guarantee that airflow and its providers are working fine together and you can install all the community providers and airflow together.

But then you are completely free to try to downgrade or upgrade any of the dependencies if they fit "requirements" independently from the constraints. That is actually the major feature of constraints that at the moment you install airflow, they are not used any more and you are free to do what you want.

However at the moment you downgrade or upgrade the dependencies, you loose guarantees that things are for sure passing the tests. and it's up to you to test it. With great powers (of upgrading and downgrading requirements) come great responsibilities (you are responsible to make sure things are working).

So you are not at all blocked. You just need to invest your time and effort if you want to add some other library that we do not have as part of our community providers to make sure that it works. But you are not blocked at all.

**Reply by Victor-Gordaliza-VM** [reply]
Thank you for the quick response!
We are going to try to modify the dependencies and test it.
